File: python/dynamic_time_wraping.py
# -*- coding: utf-8 -*-

from typing import List
from typing import Sequence
from typing import Tuple

Matrix = List[List[int]]


def get_cost_matrix(seq1: Sequence[int], seq2: Sequence[int]) -> Matrix:
    cost_matrix = [[0 for _ in seq1] for _ in seq2]

    for x2, elem2 in enumerate(seq2):
        for x1, elem1 in enumerate(seq1):
            if x2 == 0 and x1 == 0:
                penalty = 0
            elif x2 == 0:
                penalty = cost_matrix[0][x1 - 1]
            elif x1 == 0:
                penalty = cost_matrix[x2 - 1][0]
            else:
                penalty = min(cost_matrix[x2][x1 - 1],
                              cost_matrix[x2 - 1][x1 - 1],
                              cost_matrix[x2 - 1][x1])
            cost_matrix[x2][x1] = abs(elem1 - elem2) + penalty

    return cost_matrix


def get_optimal_path(
        cost_matrix: Matrix
) -> Tuple[Sequence[Tuple[int, int]], Sequence[int]]:
    optimal_path_pairs, optimal_path_costs = [], []
    x2, x1 = len(cost_matrix) - 1, len(cost_matrix[0]) - 1

    while True:
        optimal_path_pairs.append((x1, x2))
        optimal_path_costs.append(cost_matrix[x2][x1])
        if x1 == 0 and x2 == 0:
            break
        elif x1 == 0:
            x2 -= 1
        elif x2 == 0:
            x1 -= 1
        else:
            horizontal_cost = cost_matrix[x2][x1 - 1]
            diagonal_cost = cost_matrix[x2 - 1][x1 - 1]
            vertical_cost = cost_matrix[x2 - 1][x1]
            if horizontal_cost < diagonal_cost:
                if horizontal_cost < vertical_cost:
                    x1 -= 1
                else:
                    x2 -= 1
            else:
                x2 -= 1
                if diagonal_cost <= vertical_cost:
                    x1 -= 1

    return optimal_path_pairs[::-1], optimal_path_costs[::-1]


def run(
    seq1: Sequence[int], seq2: Sequence[int]
) -> Tuple[Sequence[Tuple[int, int]], Sequence[int], Matrix]:
    cost_matrix = get_cost_matrix(seq1, seq2)
    optimal_path_pairs, optimal_path_costs = get_optimal_path(cost_matrix)
    return optimal_path_pairs, optimal_path_costs, cost_matrix

File: javascript/dynamic_time_wraping.js
function makeMatrix(numRows, numCols, initialValue = 0) {
    let matrix = new Array(numRows);
    for (let i = 0; i < numRows; ++i) {
        matrix[i] = new Array(numCols)
        for (let j = 0; j < numCols; ++j) {
            matrix[i][j] = initialValue;
        }
    }
    return matrix;
}


function displayMatrix(matrix) {
    matrix.reverse();
    matrix.forEach(row => {
        row.forEach(element => process.stdout.write(`[${element}]`));
        console.log();
    });
    matrix.reverse();
}


function getCostMatrix(seq1, seq2) {
    let costMatrix = makeMatrix(seq2.length, seq1.length, 0);

    for (const [y, elem2] of seq2.entries()) {
        for (const [x, elem1] of seq1.entries()) {
            let penalty = 0;
            if (x === 0 && y === 0) {
                continue;
            } else if (x === 0) {
                penalty = costMatrix[y - 1][0];
            } else if (y === 0) {
                penalty = costMatrix[0][x - 1];
            } else {
                penalty = Math.min(costMatrix[y - 1][x], costMatrix[y - 1][x - 1], costMatrix[y][x - 1]);
            }
            costMatrix[y][x] = Math.abs(elem1-elem2) + penalty;
        }
    }

    return costMatrix;
}


function getOptimalPath(costMatrix) {
    const optimalPathCoords = [];
    const optimalPathCosts = [];
    let x = costMatrix[0].length - 1;
    let y = costMatrix.length - 1;
    
    while (true) {
        optimalPathCoords.push([x, y])
        optimalPathCosts.push(costMatrix[y][x])
        if (x === 0 && y === 0) {
            break;
        } else if (x === 0) {
            --y;
        } else if (y === 0) {
            --x;
        } else {
            const horizontalCost = costMatrix[y][x - 1];
            const diagonalCost = costMatrix[y - 1][x - 1];
            const verticalCost = costMatrix[y - 1][x];
            if (horizontalCost < diagonalCost) {
                if (horizontalCost < verticalCost) {
                    --x;
                } else {
                    --y;
                }
            } else {
                --y;
                if (diagonalCost <= verticalCost) {
                    --x;
                }
            }
        }
    }

    return [optimalPathCoords.reverse(), optimalPathCosts.reverse()];
}


function run(seq1, seq2) {
    const costMatrix = getCostMatrix(seq1, seq2);
    const [optimalPathCoords, optimalPathCosts] = getOptimalPath(costMatrix);
    return [optimalPathCoords, optimalPathCosts, costMatrix];
}


function main() {
    let seq1 = [1, 7, 3, 4, 1, 10, 5, 4, 7, 4];
    let seq2 = [1, 4, 5, 10, 9, 3, 2, 6, 8, 4];

    const [optimalPathCoords, optimalPathCosts, costMatrix] = run(seq1, seq2);
    displayMatrix(costMatrix);
    console.log("seq1 (x-axis):", seq1);
    console.log("seq2 (y-axis):", seq2);
    console.log("optimalPathCoords:", optimalPathCoords);
    console.log("optimalPathCosts:", optimalPathCosts);
}